File: day7.js
//bag class
class Bag {
  constructor(bagType, bags = []) {
    this.bagType = bagType;
    this.innerBags = bags;
  }

  getInner(index) {
    return this.innerBags[index];
  }
}
// inner bag class
class InnerBag {
  constructor(num, name) {
    this.num = num;
    this.name = name;
  }
}

//parse function for parsing out inner bags
/**
 *
 * @param {string} bagString
 */
function parseInnerBag(bagString) {
  //first char is num
  const num = parseInt(bagString.charAt(0));
  if (bagString.includes(".")) {
    bagString = bagString.slice(1, bagString.length - 5).trim();
  } else {
    bagString = bagString.slice(1, bagString.length - 4).trim();
  }
  return new InnerBag(num, bagString);
  //split on bag, grab index 0
}

/**
 *
 * @param {string} bagName
 * @param {Map} bagMap
 */
function checkForShinyGoldBag(bagName, bagMap) {
  let bag = bagMap.get(bagName);
  for (let i = 0; i < bag.innerBags.length; i++) {
    let innerBag = bag.getInner(i);
    if (innerBag.name === "shiny gold") {
      return 1;
    } else {
      if (checkForShinyGoldBag(innerBag.name, bagMap)) {
        return 1;
      }
    }
  }
  return 0;
}

/**
 *
 * @param {string} bagName
 * @param {Map} bagMap
 */
function insideBagCount(bagName, bagMap) {
  let bag = bagMap.get(bagName);
  if (bag.innerBags.length === 0) {
    return 0;
  }
  let count = 0;
  for (let i = 0; i < bag.innerBags.length; i++) {
    const num = bag.innerBags[i].num;
    count += num + num * insideBagCount(bag.innerBags[i].name, bagMap);
  }
  return count;
}

const fs = require("fs");

const input = fs.readFileSync("./input/day7.txt", "utf8").split("\r\n");

//make bagMap from rules
const bagMap = new Map();
input.forEach((rule) => {
  if (rule.includes("no other")) {
    bags = rule.split("bags");
    const outerBag = new Bag(bags[0].trim());
    bagMap.set(bags[0].trim(), outerBag);
  } else {
    bags = rule.split(",");
    const outerBagName = bags[0].split("bags")[0].trim();
    let innerBags = [];
    //parse out first inner bag from first string, loop will do the others
    innerBags.push(parseInnerBag(bags[0].split("contain")[1].trim()));
    for (let j = 1; j < bags.length; j++) {
      innerBags.push(parseInnerBag(bags[j].trim()));
    }
    bagMap.set(outerBagName, new Bag(outerBagName, innerBags));
  }
});
let count = 0;
for (let key of bagMap.keys()) {
  count += checkForShinyGoldBag(key, bagMap);
}
console.log("Part 1 :", count);
console.log("Part 2 :", insideBagCount("shiny gold", bagMap));
// console.log(checkForShinyGoldBag(bagMap.keys().next().value, bagMap, 0));

File: day7.py
print('Python')


class Bag:
    def __init__(self, bagType, bags=[]):
        self.bagType = bagType
        self.innerBags = bags

    def getInner(self, index):
        return self.innerBags[index]


class InnerBag:
    def __init__(self, num, name):
        self.num = num
        self.name = name


def parseInnerBag(bagString):
    num = int(bagString[0])
    if '.' in bagString:
        bagString = bagString[1:len(bagString) - 5].strip()
    else:
        bagString = bagString[1:len(bagString) - 4].strip()
    return InnerBag(num, bagString)


def checkForShinyGoldBag(bagName, bagMap):
    bag = bagMap[bagName]
    for i in range(len(bag.innerBags)):
        innerBag = bag.getInner(i)
        if (innerBag.name == 'shiny gold'):
            return 1
        else:
            if (checkForShinyGoldBag(innerBag.name, bagMap)):
                return 1
    return 0


def insideBagCount(bagName, bagMap):
    bag = bagMap[bagName]
    if (len(bag.innerBags) == 0):
        return 0
    count = 0
    for i in range(len(bag.innerBags)):
        num = bag.innerBags[i].num
        count += num + num * insideBagCount(bag.innerBags[i].name, bagMap)
    return count


file = open('./input/day7.txt')
file = [x.strip() for x in file.readlines()]
bagMap = {}
for rule in file:
    if "no other" in rule:
        bags = rule.split('bags')
        outerBag = Bag(bags[0].strip())
        bagMap[bags[0].strip()] = outerBag
    else:
        bags = rule.split(',')
        outerBagName = bags[0].split('bags')[0].strip()
        innerBags = []
        innerBags.append(parseInnerBag(bags[0].split('contain')[1].strip()))
        for j in range(1, len(bags)):
            innerBags.append(parseInnerBag(bags[j].strip()))
        bagMap[outerBagName] = Bag(outerBagName, innerBags)
count = 0
for key in bagMap.keys():
    count += checkForShinyGoldBag(key, bagMap)

print("Part 1 : {}".format(count))
print("Part 2 : {}".format(insideBagCount('shiny gold', bagMap)))